robot/web/stream/server.go

```diff
@@ -82,90 +47,85 @@ func (ss *streamState) Stop() {
+		}
+	}
+	return ss, nil
+}
+
+// StreamAlreadyRegisteredError indicates that a stream has a name that is already registered on
+// the stream server.
+type StreamAlreadyRegisteredError struct {
+	name string
+}
+
+func (e *StreamAlreadyRegisteredError) Error() string {
+	return fmt.Sprintf("stream %q already registered", e.name)
 }
 
-func (ss *streamServer) NewStream(config StreamConfig) (Stream, error) {
+// NewStream informs the stream server of new streams that are capable of being streamed.
+func (ss *Server) NewStream(config gostream.StreamConfig) (gostream.Stream, error) {
 	ss.mu.Lock()
 	defer ss.mu.Unlock()
 
 	if _, ok := ss.nameToStream[config.Name]; ok {
 		return nil, &StreamAlreadyRegisteredError{config.Name}
 	}
-	stream, err := NewStream(config)
+
+	stream, err := gostream.NewStream(config)
 	if err != nil {
 		return nil, err
 	}
-	if err := ss.addStream(stream); err != nil {
+
+	if err = ss.add(stream); err != nil {
 		return nil, err
 	}
-	return stream, nil
-}
-
-func (ss *streamServer) AddStream(stream Stream) error {
-	ss.mu.Lock()
-	defer ss.mu.Unlock()
-	return ss.addStream(stream)
-}
 
-func (ss *streamServer) addStream(stream Stream) error {
-	streamName := stream.Name()
-	if _, ok := ss.nameToStream[streamName]; ok {
-		return &StreamAlreadyRegisteredError{streamName}
-	}
-	ss.nameToStream[streamName] = stream
-	ss.streams = append(ss.streams, &streamState{stream: stream})
-	return nil
+	return stream, nil
 }
 
-func (ss *streamServer) Close() error {
+// ListStreams implements part of the StreamServiceServer.
+func (ss *Server) ListStreams(ctx context.Context, req *streampb.ListStreamsRequest) (*streampb.ListStreamsResponse, error) {
 	ss.mu.RLock()
-	defer ss.mu.RUnlock()
+	names := make([]string, 0, len(ss.streams))
 	for _, stream := range ss.streams {
-		stream.stream.Stop()
-		stream.activePeers = 0
-	}
-	ss.activeBackgroundWorkers.Wait()
-	return nil
-}
-
-type streamRPCServer struct {
-	streampb.UnimplementedStreamServiceServer
-	ss *streamServer
-}
-
-func (srs *streamRPCServer) ListStreams(ctx context.Context, req *streampb.ListStreamsRequest) (*streampb.ListStreamsResponse, error) {
-	srs.ss.mu.RLock()
-	names := make([]string, 0, len(srs.ss.streams))
-	for _, stream := range srs.ss.streams {
 		names = append(names, stream.stream.Name())
 	}
-	srs.ss.mu.RUnlock()
+	ss.mu.RUnlock()
 	return &streampb.ListStreamsResponse{Names: names}, nil
 }
 
-func (srs *streamRPCServer) AddStream(ctx context.Context, req *streampb.AddStreamRequest) (*streampb.AddStreamResponse, error) {
+// AddStream implements part of the StreamServiceServer.
+func (ss *Server) AddStream(ctx context.Context, req *streampb.AddStreamRequest) (*streampb.AddStreamResponse, error) {
 	pc, ok := rpc.ContextPeerConnection(ctx)
 	if !ok {
 		return nil, errors.New("can only add a stream over a WebRTC based connection")
 	}
 
 	var streamToAdd *streamState
-	for _, stream := range srs.ss.streams {
+	for _, stream := range ss.streams {
```

Comment:
I see that `AddStream` doesn't lock `mu` before accessing `ss.streams`. Is `streams` not protected by `mu`? if not, could we please move all `Server` struct members protected by `mu` below `mu` & everything not protected by it above `mu`?

Reply:
This is a bit of a mess. Now that you've highlighted this, I need to restrain myself from just doing all of this over.

Notably there's a [write-write race](https://github.com/viamrobotics/rdk/pull/3699/files#diff-b664e7f5f24e04be1ed075bf7d0f4209eaf9fc88f34eb7acc987ad1693857223R266) if multiple threads call `Server.Stop` at the same time.